libr/include/r_util/r_rbtree.h

```diff
@@ -1,75 +1,90 @@
+/* radare - BSD 3 Clause License - Copyright 2017 - MaskRay */
+
 #ifndef R2_RBTREE_H
 #define R2_RBTREE_H
 
 #include <limits.h>
 #include <stdbool.h>
+#include <stddef.h>
 
 #include "r_list.h"
 
+#ifndef container_of
+# ifdef _MSC_VER
+#  define container_of(ptr, type, member) ((type *)((char *)(ptr) - offsetof(type, member)))
+# else
+#  define container_of(ptr, type, member) ((type *)((char *)(__typeof__(((type *)0)->member) *){ptr} - offsetof(type, member)))
```

Comment:
`__typeof__` is a gnu extension, is there a desire for this to be ISO compliant? we try to shoot for that in r2land when possible.

Reply:
see https://stackoverflow.com/a/10269766 for a suggestion. I haven't checked for correctness.

Reply:
I took the second answer.

You gain type safety with the `__typeof__` macro definition: if `ptr` is not a pointer of type `struct type::member`, it does not compile.

Reply:
How about changing the check to `__GNUC__` - checking if compiler claims to support the gnu extensions instead of one of the compilers that doesn't? From the stackoverflow it seems that additional code in second version is for extra typechecking and is completely optional.  If it causes problems it will be easy to switch to first version. Targeting ISO compliance doesn't mean that additional extensions can't be used for additional error checking when available assuming they are correctly detected and do not increase the maintaince work.

Reply:
Use `_MSC_VER` to make it striking that it is not a GNU specific extension (it's actually widely supported) and remove it when MSVC is smarter.